snapshotter/utils/generic_worker.py

```diff
@@ -67,11 +92,135 @@ def __init__(self, name, **kwargs):
         self._commit_payload_exchange = (
             f'{settings.rabbitmq.setup.commit_payload.exchange}:{settings.namespace}'
         )
+        self._event_detector_exchange = f'{settings.rabbitmq.setup.event_detector.exchange}:{settings.namespace}'
+        self._event_detector_routing_key_prefix = f'powerloom-event-detector:{settings.namespace}:{settings.instance_id}.'
         self._commit_payload_routing_key = (
             f'powerloom-backend-commit-payload:{settings.namespace}:{settings.instance_id}.Data'
         )
         self._initialized = False
 
+    def _signal_handler(self, signum, frame):
+        if signum in [SIGINT, SIGTERM, SIGQUIT]:
+            self._core_rmq_consumer.cancel()
+
+    @retry(
+        wait=wait_random_exponential(multiplier=1, max=10),
+        stop=stop_after_attempt(5),
+        retry=tenacity.retry_if_not_exception_type(httpx.HTTPStatusError),
+        after=web3_storage_retry_state_callback,
+    )
+    async def _upload_web3_storage(self, snapshot: bytes):
+        web3_storage_settings = settings.web3storage
+        files = {'file': snapshot}
+        r = await self._web3_storage_upload_client.post(
+            url=f'{web3_storage_settings.url}{web3_storage_settings.upload_url_suffix}',
+            files=files,
+        )
+        r.raise_for_status()
+        resp = r.json()
+        self._logger.info('Uploaded snapshot to web3 storage: {} | Response: {}', snapshot, resp)
+
+    async def _commit_payload(
+            self,
+            type_: str,
+            _ipfs_writer_client: AsyncIPFSClient,
+            project_id: str,
+            epoch: Union[
+                PowerloomSnapshotProcessMessage,
+                PowerloomSnapshotSubmittedMessage,
+                PowerloomCalculateAggregateMessage,
+            ],
+            snapshot: Union[BaseModel, AggregateBase],
+            storage_flag: bool,
+    ):
+        # payload commit sequence begins
+        # upload to IPFS
+        snapshot_json = json.dumps(snapshot.dict(by_alias=True), sort_keys=True, separators=(',', ':'))
+        snapshot_bytes = snapshot_json.encode('utf-8')
+        try:
+            snapshot_cid = await _ipfs_writer_client.add_bytes(snapshot_bytes)
```

Comment:
We definitely need so sort of retry mechanism for IPFS uploads as well, either we can do that here or in IPFS client library. Better to do it here IMO, that'll keep the library clean.